crypt-yescrypt.c

```diff
@@ -0,0 +1,108 @@
+/* Copyright (C) 2018 vt@altlinux.org
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include "crypt-port.h"
+#include "crypt-private.h"
+#include "byteorder.h"
+
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+#include "yescrypt.h"
+
+#if INCLUDE_yescrypt
+
+/*
+ * As OUTPUT is initialized with a failure token before gensalt_yescrypt_rn
+ * is called, in case of an error we could just set an appropriate errno
+ * and return.
+ * Since O_SIZE is guaranteed to be greater than 2, we may fill OUTPUT
+ * with a short failure token when need.
+ */
+void
+gensalt_yescrypt_rn(unsigned long count,
+                     const uint8_t *rbytes, size_t nrbytes,
+                     uint8_t *output, size_t o_size)
+{
+	/* Use one of recommended parameter sets as the 'low default'. */
+	yescrypt_params_t params = { .flags = YESCRYPT_DEFAULTS,
+		.N = 4096, .r = 32, .p = 1 };
+
+	if (count) {
+		/*
+		 * `1 << (count - 1)` is MiB usage in range of 1MiB..1GiB,
+		 * thus, count is in range of 1..11
+		 */
+		if (count <= 2) {
+			params.r = 8; /* N in 1KiB */
+			params.N = 512ULL << count;
+		} else if (count <= 11) {
+			params.r = 32; /* N in 4KiB */
+			params.N = 128ULL << count;
+		} else {
+			errno = EINVAL;
+			return;
+		}
+	}
+
+	if (!yescrypt_encode_params_r(&params, rbytes, nrbytes, output, o_size)) {
+		/*
```

Comment:
`make_failure_token()` from `crypt.c` could be used here.  It just needs some small tweaks to the general code-base.

Reply:
Another (and the better idea instead of my first sugestion) would be to allocate a temporary buffer with `o_size` but not more than max. needed length and strncpy it to output after success.

Reply:
Why mess with allocation and copying, if I just can write 2-3 char? My choice is better in terms of performance and simplicity of code. I also don't need token alternation functionality from `make_failure_token`. You also have read out of bounds in `make_failure_token`.